Add support for zstd compression

abitofevrything opened the pull request:
# Description

Add support for zstd-stream transport compression.

Drafted for now so we can determine whether adding a dependency on another package for this is worth it, especially as it is an unofficial third-party package based on FFI, and not portable.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] Ran `dart analyze` or `make analyze` and fixed all issues
- [x] Ran `dart format --set-exit-if-changed -l 160 ./lib` or `make format` and fixed all issues
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

abitofevrything committed:
Add support for zstd compression

abitofevrything committed:
Add comment

l7ssha commented on lib/src/gateway/shard_runner.dart:
```diff
@@ -465,6 +467,9 @@ Stream<dynamic> decompressTransport(Stream<List<int>> raw) {
   });
 }
 
+// See https://github.com/instantiations/es_compression/issues/52 for why the isNotEmpty check is needed.
+Stream<dynamic> decompressZStdTransport(Stream<List<int>> raw) => raw.transform(zstd.decoder).where((message) => message.isNotEmpty);
```
is isNotEmpty still needed? https://github.com/instantiations/es_compression/issues/52#issuecomment-2336428316

abitofevrything committed:
Prefer hand written zstd decoder

abitofevrything committed:
Switch to ETF+ZStd by default

abitofevrything committed:
Don't send empty output

abitofevrything committed:
Decode each gateway message as a whole unit

abitofevrything committed:
Reduce integration test concurrency

abitofevrything committed:
Fix empty gateway messages

I suspect these were simply initialisation frames that didn't produce any output.

abitofevrything committed:
Correctly set the output buffer size

abitofevrything committed:
Remove bad sanity check

It's better we crash, it means something actually went wrong in the decompression
